impeller/renderer/backend/vulkan/fence_waiter_vk.cc

```diff
@@ -51,59 +93,83 @@ void FenceWaiterVK::Main() {
+    if (terminate) {
       break;
     }
-    if (wait_set->empty()) {
+
+    // Check if the context had died in the meantime.
+    auto device_holder = device_holder_.lock();
+    if (!device_holder) {
+      break;
+    }
+
+    const auto& device = device_holder->GetDevice();
+
+    // Wait for one or more fences to be signaled. Any additional fences added
+    // to the waiter will be serviced in the next pass. If a fence that is going
+    // to be signaled at an abnormally long deadline is the only one in the set,
+    // a timeout will bail out the wait.
+    auto fences = GetFencesForWaitSet(wait_set);
+    if (fences.empty()) {
       continue;
     }
 
-    auto result = device_holder->GetDevice().waitForFences(
-        wait_set->size(),                        // fences count
-        wait_set->data(),                        // fences
+    auto result = device.waitForFences(
+        fences.size(),                           // fences count
+        fences.data(),                           // fences
         false,                                   // wait for all
         std::chrono::nanoseconds{100ms}.count()  // timeout (ns)
     );
     if (!(result == vk::Result::eSuccess || result == vk::Result::eTimeout)) {
+      VALIDATION_LOG << "Fence waiter encountered an unexpected error. Tearing "
+                        "down the waiter thread.";
       break;
     }
-  }
-}
 
-std::optional<std::vector<vk::Fence>> FenceWaiterVK::TrimAndCreateWaitSetLocked(
-    const std::shared_ptr<DeviceHolder>& device_holder) {
-  if (terminate_) {
-    return std::nullopt;
-  }
-  TRACE_EVENT0("impeller", "TrimFences");
-  std::vector<vk::Fence> fences;
-  fences.reserve(wait_set_.size());
-  for (auto it = wait_set_.begin(); it != wait_set_.end();) {
-    switch (device_holder->GetDevice().getFenceStatus(it->first->Get())) {
-      case vk::Result::eSuccess:  // Signalled.
-        it->second();
-        it = wait_set_.erase(it);
-        break;
-      case vk::Result::eNotReady:  // Un-signalled.
-        fences.push_back(it->first->Get());
-        it++;
-        break;
-      default:
-        return std::nullopt;
+    // One or more fences have been signaled. Find out which ones and update
+    // their signaled statuses.
+    {
+      TRACE_EVENT0("impeller", "CheckFenceStatus");
+      for (auto& entry : wait_set) {
+        entry->UpdateSignalledStatus(device);
+      }
+      wait_set.clear();
+    }
+
+    // Quickly acquire the wait set lock and erase signaled entries. Make sure
+    // the mutex is unlocked before calling the destructors of the erased
+    // entries. These might touch allocators.
+    WaitSet erased_entries;
+    {
+      static auto is_signalled = [](const auto& entry) {
+        return entry->IsSignalled();
+      };
+      std::scoped_lock lock(wait_set_mutex_);
+      std::copy_if(wait_set_.begin(), wait_set_.end(),
+                   std::back_inserter(erased_entries), is_signalled);
+      wait_set_.erase(
```

Comment:
Aren't you still calling the callback in the while inside the lock?  The last reference deleted should be here (inside of a lock).  On the other PR you said you wanted to remove the callback outside of the lock.

Reply:
Probably the easiest fix is to make UpdateSignalStatus call the callback.

This code would probably be easier to follow if it was converted to use `std::unique_ptr` then holding raw pointers on the stack in `wait_set`.

Reply:
The entries are shared pointers to `WaitSetEntry`. These have been copied over to the erased_entries set. That set is cleared outside the critical section. The scoped cleanup closures get dropped in `WaitSetEntry` dtor.

Reply:
Ahh, I missed `erased_entries`.  LGTM.